metadetect/util.py

```diff
@@ -127,6 +128,218 @@ def get_ratio_error(a, b, var_a, var_b, cov_ab):
+        # the calculation of the response
+
+        noise_replacer_config = NoiseReplacerConfig()
+        footprints = {
+            source.getId(): (source.getParent(), source.getFootprint())
+            for source in sources
+        }
+
+        # This constructor will replace all detected pixels with noise in the
+        # image
+        self.replacer = NoiseReplacer(
+            noise_replacer_config,
+            exposure=exposure,
+            footprints=footprints,
+            noiseImage=noise_image,
+        )
+
+    @contextmanager
+    def sourceInserted(self, source_id):
+        """
+        Context manager to insert a source
+
+        with replacer.sourceInserted(source_id):
+            # do something with exposure
+        """
+        self.insertSource(source_id)
+
+        try:
+            yield self
+        finally:
+            self.removeSource(source_id)
+
+    def insertSource(self, source_id):
+        """
+        Insert a source
+        """
+        self.replacer.insertSource(source_id)
+
+    def removeSource(self, source_id):
+        """
+        Remove a source
+        """
+        self.replacer.removeSource(source_id)
+
+    def end(self):
+        self.replacer.end()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        self.end()
+
+
+class MultibandNoiseReplacer(object):
+    """
+    noise replacer that works on multiple bands
+
+    Parameters
+    ----------
+    mbexp: lsst.afw.image.MultibandExposure
+        The data
+    sources: lsst.afw.table.SourceCatalog
+        Catalog of sources
+
+    Examples
+    --------
+    with MultibandNoiseReplacer(mbexp=mbexp, sources=sources) as replacer:
+        with replacer.insertSource(source_id):
+            # do something with exposures
+    """
+    def __init__(self, mbexp, sources):
+        self.mbexp = mbexp
+        self.sources = sources
+        self._set_noise_replacers()
+
+    @contextmanager
+    def sourceInserted(self, source_id):
+        """
+        Context manager to insert a source
+
+        with replacer.sourceInserted(source_id):
+            # do something with exposures
+        """
+        self.insertSource(source_id)
+
+        try:
+            # usually won't use this yielded value
+            yield self.mbexp
+        finally:
+            self.removeSource(source_id)
+
+    def insertSource(self, source_id):
+        """
+        Insert a source
+        """
+        for replacer in self.noise_replacers:
+            replacer.insertSource(source_id)
+
+    def removeSource(self, source_id):
+        """
+        Remove a source
+        """
+        for replacer in self.noise_replacers:
+            replacer.removeSource(source_id)
+
+    def end(self):
+        """
+        end the noise replacment.  Called automatically upon leaving the
+        context manager
+        """
+        # we do nothing because our replacers are context managers
```

Comment:
This is not quite right I think. You only call exit if you use the replaces with a `with` statementent. That is not happening in the code snippet above. 

I think the right way to do this is to chain the context managers using `contextlib.ExitStack`.

Reply:
correct, I didn't know about ExitStack

Note this class is used in my shredder branch which branched off this branch